### ISSUE TITLE
Support dfu files

### ISSUE THREAD

**Issue by publicarray** [pine64/pinecil-firmware-updater#17]
The new [release candidate 2.16](https://github.com/Ralim/IronOS/releases/tag/v2.16-rc)  is shipping `dfu` files.

**Comment by gamelaster** [comment]
@publicarray thanks for letting me know!
Although, I'm wondering if pushing RCs into updater is good idea.

**Comment by publicarray** [comment]
Yea this is more of a heads up for the next release. 
I also like to see support for it when manually choosing a firmware file (to potentially install the RC)

**Comment by Johnex** [comment]
The dfu-util needs to also be updated to version 0.11 as the old version can't handle the new headers, thanks at @Ralim for hinting about it.

https://github.com/Ralim/IronOS/issues/1070

**Comment by tormodvolden** [comment]
Older dfu-util like 0.10 or 0.9 should work fine with these DfuSe files, you just need to add `-a 0` , which became optional in 0.11.

**Comment by gamelaster** [comment]
@Johnex thanks for letting me know, and @tormodvolden this is good news, since I will not need to build the latest version 😁 (also this is good for Linux distributions which don't have latest dfu-util)

**Comment by tormodvolden** [comment]
Unfortunately, many Linux distributions are still on 0.9, which lacks the tweaks needed for working around the bugs in the GD32VF103 chip bootloader.

**Comment by tormodvolden** [comment]
I recommend all users to use 0.11, and strongly recommends anyone redistributing it (!) to ship 0.11, because the diagnostics output has been improved which helps debugging issues.

**Comment by gamelaster** [comment]
@tormodvolden well, on some OSes it can be problem to upgrade to latest, this is why I shipped binaries with pre-compiled latest versions

**Comment by publicarray** [comment]
With almost any small community/open-source project with no commercial entity or roadmap there is no ETA on releases.

**Comment by eikaramba** [comment]
stable version is released now :)

**Comment by gamelaster** [comment]
@eikaramba yes, I'm aware of that and I will try to do it in following days 😊

**Comment by gejanssen** [comment]
For the guys unable to wait, I updated my firmware (2.16) using the following command:

sudo dfu-util -d 28e9:0189 -a 0 -D Pinecil_EN.dfu

mind that the previous version was updated with the extra command: -s 0x08000000 and the help page of dfu-util 0.9 says;

  -s --dfuse-address <address>	ST DfuSe mode, specify target address for
				raw file download or upload. Not applicable for
				DfuSe file (.dfu) downloads

**Comment by tormodvolden** [comment]
In general: Use `-s` (= `--dfuse-address`) with raw binary files (with or without a DFU suffix concatenated to the file contents). Such files typically have a .bin filename extension, but sometimes people use the .dfu filename extension if the file include a DFU suffix (inside the file).
Don't use `-s` with DfuSe files because they are specially structured files with the load addresses embedded. The DfuSe files always have a DFU suffix inside. Part of the DFU suffix tell if the file is a DfuSe file (DFU file version 1.1a) or a normal DFU file (version 1.0 or 1.1).

I suspect some people might confuse _.dfu filename extension_ with _DFU suffix_ (which is part of the file contents)...

When using DfuSe files, the `-a` option was needed up to dfu-util 0.10, but it can be skipped from dfu-util 0.11 on (if all images inside the file should be applied, otherwise -a is used to select one).

**Comment by tormodvolden** [comment]
Note also that the DFU suffix normally contains the device identifier, so the `-d` option is not needed (unless you want the override the suffix values).

**Comment by Johnex** [comment]
With 0.11 and later you just have to do ```dfu-util  Pinecil_EN.dfu``` and nothing else. If someone mistakenly changes the file extension from bin to dfu, the command will just fail anyway.

@gamelaster Any update on progress?

**Comment by tormodvolden** [comment]
> With 0.11 and later you just have to do `dfu-util Pinecil_EN.dfu` and nothing else. If someone mistakenly changes the file extension from bin to dfu, the command will just fail anyway.

Unfortunately wrong on both accounts. You need the -D option to tell that you want to "download" to the device. If you change the filename extension it works exactly the same because dfu-util doesn't look at the filename extension, only at the file contents.

**Comment by Johnex** [comment]
Yes sorry, it needs the -D, i didn't copy the command correctly, i updated above. The second part i mean, if you take a bin file, with bin contents, and change it to .dfu, dfu-util will not flash the wrong file. It will tell you first of all that you need to select which partition to flash to and stop.

**Comment by choekstr** [comment]
I tried using the .10 version of dfu-util (win10 btw) with the full string including hex address but it just errored out saying it reported a bad page size for it's internal memory.  It was easy to get the updated .11 version of dfu-util (just need the new .exe and the new libusb-1.0.dll which has changed despite being the same filename!).
Once I had the updated dfu-util then the dfu-util -D Pinecil_EN.dfu worked perfectly from the cmdline (either powershell or cmd seemed to work when I tested both)

Thanks for the help all, this is what I came looking for while we wait for an updated version of the gui and found exactly what I wanted.  Brilliant community!

**Comment by tormodvolden** [comment]
Yes, there was a bug in the Windows 64-bit build of dfu-util 0.10 that affected DFU suffix parsing. Maybe this was the reason.

**Comment by moriel5** [comment]
Is there any reason why there haven't yet been any new commits to the `master` branch in regards to this? Since I would have expected initial untested code to exist, with the obvious caveats of not being deployed as a stable release.

**Comment by Johnex** [comment]
The new DFU changes were made in the release candidate of IronOS that came out more than 2 months ago. We asked the owner of this project @gamelaster if there would be an update here and he said "when the stable version will be released". For me this is quite funny since in all this time there could have been a release candidate here of the updater in parallel, instead of waiting for a release and then taking multiple weeks after to make an update.

**Comment by moriel5** [comment]
@Johnex Agreed, this is strange, especially since the stable release has already been released and still naught has been done.

Also, and this is probably out of scope, however with the PineTime's updates also being packaged as DFUs, it would be nice if this installer could be extended to support those, as well as possibly additional Pine64 devices.

**Comment by gamelaster** [comment]
@Johnex I wanted to make support for DFU firmware files even before IronOS stable release, although, I'm maintaining all of this in my free time, and I have a lot of another things to do. So, I'm very welcome to pull requests or any help around this project, so feel free to contribute :-)
Anyway, as I said, this is planned as to be done, and I will try to implement this as soon as possible (together with other issues I was reading few days ago).

@moriel5 It's not out of scope, actually, the chance is very high that this application will become updater also for PineTime and other PINE64 products (or maybe also other?), but when all this will happen is at the moment unknown.

**Comment by Johnex** [comment]
@gamelaster If that was the case, why not say it when we asked the first time, and not "when there is a stable release"?

![ObamaWtfGIF](https://user-images.githubusercontent.com/7826974/149426934-20d3cd6a-3d86-4f76-b26f-28943f98eedf.gif)

**Comment by Ralim** [comment]
Dude, leave the guy alone. He's already said hes aware and he will get to it when he can. Hes not _paid_ to work on any of this, so its only when time and motivation are around. If you need it you are always welcome to submit a PR, this is an open source project :).

**Comment by moriel5** [comment]
@gamelaster Thanks for the confirmation, that does make sense.

And I'm not trying to bash, since I can understand what you are saying and respect it, however I think that this misunderstanding was caused by lack of linguistic transparency, something that easily happens with everyone (as in, at least mentioning the reason for delay somewhere like the Pine64 forums, though you are not really required to do so).

**Comment by Johnex** [comment]
@Ralim I don't need it, i just thought it was a strange way to go about the whole thing. Most people that DO need it are probably not experienced enough to make a pull request, but sure, all the noobies here can make a pull request and contribute 😆

**Comment by choekstr** [comment]
It's just as easy to run the 1 rev older version while people wait.  Those noobies aren't going to be pushing for cutting edge anyway and if they are really needing/wanting the newest version, they can read a bit and google a bit and eventually find steps to upgrade manually.  It'll come, just be patient and let gamelaster work on it when he gets a chance.

**Comment by Sokoloft** [comment]
> For the guys unable to wait, I updated my firmware (2.16) using the following command:
> 
> sudo dfu-util -d 28e9:0189 -a 0 -D Pinecil_EN.dfu


Same command worked for me. Thanks.

**Comment by sixtyfive** [comment]
Thanks, @gejanssen, for the info about 

> -s --dfuse-address

Using

```bash
sudo dfu-util -d 28e9:0189 -a 0 -D Pinecil_EN.dfu
```

I was also able to update to 2.16.
